src/smtp.rs

```diff
@@ -510,5 +515,155 @@ pub(crate) async fn send_smtp_messages(context: &Context, connection: &mut Smtp)
             success = false;
         }
     }
+
+    loop {
+        match send_mdn(context, connection).await {
+            Err(err) => {
+                info!(context, "Failed to send MDNs over SMTP: {:#}.", err);
+                success = false;
+                break;
+            }
+            Ok(false) => {
+                break;
+            }
+            Ok(true) => {}
+        }
+    }
     Ok(success)
 }
+
+/// Tries to send MDN for message `msg_id` to `contact_id`.
+///
+/// Attempts to aggregate additional MDNs for `contact_id` into sent MDN.
+///
+/// On failure returns an error without removing any `smtp_mdns` entries, the caller is responsible
+/// for removing the corresponding entry to prevent endless loop in case the entry is invalid, e.g.
+/// points to non-existent message or contact.
+async fn send_mdn_msg_id(
+    context: &Context,
+    msg_id: MsgId,
+    contact_id: ContactId,
+    smtp: &mut Smtp,
+) -> Result<()> {
+    let contact = Contact::load_from_db(context, contact_id).await?;
+    if contact.is_blocked() {
+        return Err(format_err!("Contact is blocked"));
+    }
+
+    // Try to aggregate additional MDNs into this MDN.
+    let additional_msg_ids: Vec<MsgId> = context
+        .sql
+        .query_map(
+            "SELECT msg_id
+             FROM smtp_mdns
+             WHERE from_id=? AND msg_id!=?",
+            paramsv![contact_id, msg_id],
+            |row| {
+                let msg_id: MsgId = row.get(0)?;
+                Ok(msg_id)
+            },
+            |rows| rows.collect::<Result<Vec<_>, _>>().map_err(Into::into),
+        )
+        .await?;
+
+    // Load corresponding RFC724 message IDs
+    let mut additional_rfc724_mids = Vec::new();
+    for msg_id in &additional_msg_ids {
+        if let Ok(Message { rfc724_mid, .. }) = Message::load_from_db(context, *msg_id).await {
```

Comment:
I added `rfc724_mid` column to `smtp_mdns` table to avoid dealing with cases where message is deleted from `msgs` table and we cannot load `rfc724_mid` from there.